cirq-core/cirq/transformers/routing/line_initial_mapper.py

```diff
@@ -0,0 +1,237 @@
+        logical qubits given by the circuit graph onto one long line of physical qubits on the
+        device, greedily maximizing each physical qubit's degree.
+        If this mapping cannot be completed as one long line of qubits in the circuit graph mapped
+        to qubits in the device graph, the line can be split as several line segments and then we:
+            (i)   Map first line segment.
+            (ii)  Find another high degree vertex in G near the center.
+            (iii) Map the second line segment
+            (iv)  etc.
+
+        Args:
+            circuit: the input circuit with logical qubits
+
+        Returns:
+            a dictionary that maps logical qubits in the circuit (keys) to physical qubits on the
+            device (values).
+        """
+        return self._initial_mapping(circuit.freeze())
+
+    @_compat.cached_method
+    def _initial_mapping(self, circuit: 'cirq.FrozenCircuit') -> Dict['cirq.Qid', 'cirq.Qid']:
+        """Maps disjoint lines of logical qubits onto lines of physical qubits.
+
+        Helper for 'initial_mapping' that takes a (hashable) frozen circuit to cache the result.
+
+        Args:
+            circuit: the input circuit with logical qubits
+
+        Returns:
+            a dictionary that maps logical qubits in the circuit (keys) to physical qubits on the
+            device (values).
+        """
+        qubit_map: Dict['cirq.Qid', 'cirq.Qid'] = {}
+        circuit_graph = self._make_circuit_graph(circuit)
+        physical_center = nx.center(self.device_graph)[0]
+
+        def next_physical(current_physical: 'cirq.Qid') -> 'cirq.Qid':
+            # use current physical if last logical line ended before mapping to it.
+            if self.device_graph.nodes[current_physical]["mapped"] is False:
+                return current_physical
+            # else greedily map to highest degree neighbor that that is available
+            sorted_neighbors = sorted(
+                self.device_graph.neighbors(current_physical),
+                key=lambda x: self.device_graph.degree(x),
+            )
+            for neighbor in reversed(sorted_neighbors):
+                if self.device_graph.nodes[neighbor]["mapped"] is False:
+                    return neighbor
+            # if cannot map onto one long line of physical qubits, then break down into multiple
+            # small lines by finding nearest available qubit to the physical center
+            return self._closest_unmapped_qubit(physical_center)
+
+        def next_logical(current_logical: 'cirq.Qid') -> Optional['cirq.Qid']:
+            for neighbor in circuit_graph.neighbors(current_logical):
+                if circuit_graph.nodes[neighbor]["mapped"] is False:
+                    return neighbor
+            return None
+
+        for pq in self.device_graph.nodes:
+            self.device_graph.nodes[pq]["mapped"] = False
```

Comment:
Can we use a set `mapped = {}` instead, and query "if current_physical in mapped" instead of `self.device_graph.nodes[current_physical]["mapped"] is False` ? 

This would help not increase the memory footprint of the entire class (by adding attributes to the `self.device_graph` instead of using a local data structure that gets destroyed after going out of scope).